### ISSUE TITLE
LightGBM produces identical leaf values for majority of the leaves

### ISSUE THREAD

**Issue by naropa7** [microsoft/LightGBM#6027]
## Description
Hi! I noticed the most of the leaf values in the trees share identical values. I am using `trees_to_dataframe()` function to inspect the leave values. The identical leaf values either equal to the minimal leaf value or the maximum leaf value of the tree without exception. I'm curious what is the cause of this behavior. These leaves have different numbers of samples and weights. I would expect some differences in their leaf values. I've tested different datasets and saw the same behavior. Thank you!

## Reproducible example
Below is the params I'm using and the results I got:

```
{'boosting': 'gbdt',
   'metric': 'rmse',
   'bagging_freq': 1,
   'verbosity': -1,
   'n_jobs': -1,
   'early_stopping_round': 10,
   'num_rounds': 10000,
   'seed': 0,
   'deterministic': True,
   'force_col_wise': True,
   'bagging_fraction': 0.9745239331700798,
   'cat_l2': 17.0,
   'cat_smooth': 285.0,
   'feature_fraction': 0.5012123832040147,
   'learning_rate': 0.07649878768592239,
   'max_cat_threshold': 74,
   'max_depth': 5,
   'min_data_in_leaf': 119,
   'min_data_per_group': 136,
   'num_leaves': 56,
   'max_delta_step': 0.43872434313624487}
```

<img width="1208" alt="Screenshot 2023-08-09 at 5 06 19 PM" src="https://github.com/microsoft/LightGBM/assets/141875875/ee7d2b9d-55d1-438a-aac1-9be4ed7b8b2a">

## Environment info

LightGBM version or commit hash:
LightGBM '3.3.4'

Command(s) you used to install LightGBM

```shell
pip install lightgbm

```

**Comment by jameslamb** [comment]
> I've tested different datasets and saw the same behavior

Are you able to provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), some self-contained code we could run which replicates this?

For example, like this

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=10_000)
dtrain = lgb.Dataset(X, label=y)
bst = lgb.train(train_set=dtrain, params={"objective": "rmse"})
```

You're right to note that this is unexpected, but it's not impossible. I suspect this outcome is highly dependent on your data, for example on the distribution of the target variable.

It's going to be difficult to investigate this given only the parameters and a screenshot of that dataframe representation of the Booster.

**Comment by jameslamb** [comment]
Also, is it easy for you to test with `lightgbm` v4.0.0? That release contains a LOT of bugfixes and other changes relative to the version you're using (v3.3.4). Perhaps the issue is already fixed in v4.0.0.

**Comment by naropa7** [comment]
Hi! Thank you for the quick response! I just tested on V4.0.0 and saw the same behavior. Unfortunately I'm working on private datasets so I couldn't share here. But let me see if I can replicate this behavior on public datasets and get back to you. 

I think your suspicion is correct though. The datasets I'm working on have very noisy target distributions. I'm curious what ensures the identical leave values only equal to the min or max leaf values but never in-between. 

Based on this [doc](https://lightgbm.readthedocs.io/en/latest/_modules/lightgbm/basic.html#Booster.trees_to_dataframe), the `value` in a leaf is the predicted value for this leaf node, multiplied by the learning rate. Is the "predicted value" here simply the average value of the labels among the samples within the same leaf? Should I expect two leaves with different samples/weights and average labels to have different leaf values?

**Comment by jameslamb** [comment]
>  Is the "predicted value" here simply the average value of the labels among the samples within the same leaf?

The value in a leaf node is the the one that minimizes the aggregate loss for all examples that fall into that leaf node.

To create a prediction for one sample, LightGBM passes that sample through each tree, then sums the values from all leaf nodes the sample falls into.

So those values you see in the `value` column are in terms of the gradients (which you can think of conceptually as model residuals), not the actual target. So for example, the predicted value for a leaf node being `1.1` does not mean that the average of all the values of the label for observations in that leaf node.

The only time that would be true is if you've passed the parameter `boost_from_average=False`, and if you're looking at the structure of the first tree in the ensemble.

> Should I expect two leaves with different samples/weights and average labels to have different leaf values?

You should expect that two leaves *with the same parent node* have different leaf values. If they didn't, then the gain for that split would be 0 and it might as well not have been made.

If you find a situation like that, that sounds like a bug in LightGBM to me unless you are using forced splits and forcing LightGBM to split a certain way even if that way generates 0 gain.

It's technically possible, though unlikely in my experience, for leaf nodes with different parents to happen to have identical predicted values.

**Comment by naropa7** [comment]
Thank you for the thorough explanation! It was very helpful. I was able to recalculate the leave value and matched the output above. The cause was the combination of `max_delta_step` and `learning_rate`. Since both are set to relatively small values in my dataset, and I have a pretty noisy label distribution, a lot of the leave values exceeded `learning_rate * max_delta_step = 0.034` in the example above. Thus, their leave values were clipped at [mean labels - 0.034, mean labels + 0.034], which are the [1.112882, 1.180005] values in the screenshot above. 

I believe the package is working as intended. Please feel free to close this thread. Appreciate your team maintaining such a great package!

**Comment by jameslamb** [comment]
Ahhhh interesting! I did not notice `max_delta_step` in the parameters you provided. That makes sense to me.

Thanks for using LightGBM and for being so responsive. Come back any time 👋🏻
